### PR TITLE
AWS Lambda SDK: Improved sampling algorithm

### PR DESCRIPTION
**Pull request by medikoo**
- Recognition of ALB HTTP event
- Improvement to sampling algorithm according to internal agreements

### CODE REVIEW

File: node/packages/aws-lambda-sdk/instrument/index.js [hunk 1]
```diff
@@ -98,15 +99,27 @@ const reportResponse = async (response, context, endTime) => {
   await sendTelemetry('request-response', payloadBuffer);
 };
 
+let requestCounter = 0;
+let lastCounterResetTime = 0;
+
 const reportTrace = ({ isErrorOutcome }) => {
   // Sample out 80% of traces in production mode if no warning or error event was reported
```
Comment (Danwakeem) [inline node/packages/aws-lambda-sdk/instrument/index.js]: I think I am following looking at the code but not 100% sure. I think an updated comment would help 🙏 

Does this sound right?

```
/**
 * This function determines if a trace should be sampled out or not
 * 
 * 
 * Traces will be sampled out when
 *    Trace contains Errors or Warnings
 *    Dev Mode = enabled
 *    Debug Mode = enabled
 *    Cold start
 *    Lambda is an API and has received >2 requests within one second
 * 
 * Given none of the conditions above are met we will output 90% of traces
*/
```

~~I think I am missing the part about `isApiEvent`~~ 🤔 Updated to include the API event exclusion

File: node/packages/aws-lambda-sdk/instrument/index.js [hunk 2]
```diff
@@ -98,15 +99,27 @@ const reportResponse = async (response, context, endTime) => {
...
+      if (serverlessSdk._isDebugMode) return false;
+      if (serverlessSdk._isDevMode) return false;
+      if (capturedEvents.some(({ name }) => alertEventNames.has(name))) return false;
+      const currentTime = Date.now();
+      if (currentTime - lastCounterResetTime > 1000) {
+        lastCounterResetTime = currentTime;
+        requestCounter = 0;
+      }
+      ++requestCounter;
+      if (isApiEvent()) return requestCounter > 2;
```
Comment (Danwakeem) [inline node/packages/aws-lambda-sdk/instrument/index.js:126]: Does this mean that after the first 2 API requests all other requests within 1 second will be reported?

**Review by Danwakeem — COMMENTED** [review]
This is clever, I like it 😎 But I think the logic is complicated enough it might be worth updating the comment.

File: node/packages/aws-lambda-sdk/instrument/index.js (hunk 2 above)
Comment (medikoo) [reply node/packages/aws-lambda-sdk/instrument/index.js:126]: After 2 API requests they won't be reported (`true` resolves to sampled out)

Comment (Danwakeem) [reply node/packages/aws-lambda-sdk/instrument/index.js:126]: Ahh okay got I had it backwards. Thanks for the clarification 👍

File: node/packages/aws-lambda-sdk/instrument/index.js (hunk 1 above)
Comment (medikoo) [reply node/packages/aws-lambda-sdk/instrument/index.js]: Good point, I forgot to update this comment. I've updated it now

**Review by Danwakeem — APPROVED** [review]
Looks great :shipit:

**Pull request merged**